File: packages/python/src/shared_platform/webhooks/signature.py
"""Webhook signature utilities."""
from __future__ import annotations

import hashlib
import hmac
import time
from typing import Optional

from .exceptions import InvalidSignatureError


def generate_signature(
    payload: str | bytes,
    secret: str,
    timestamp: Optional[int] = None,
) -> tuple[str, int]:
    """
    Generate HMAC-SHA256 signature for webhook payload.

    Args:
        payload: The raw JSON payload (string or bytes)
        secret: The webhook secret key
        timestamp: Optional Unix timestamp (defaults to current time)

    Returns:
        Tuple of (signature, timestamp)
    """
    if timestamp is None:
        timestamp = int(time.time())

    if isinstance(payload, str):
        payload = payload.encode("utf-8")

    # Create signed payload: timestamp.payload
    signed_payload = f"{timestamp}.".encode("utf-8") + payload

    # Generate HMAC-SHA256 signature
    signature = hmac.new(
        secret.encode("utf-8"),
        signed_payload,
        hashlib.sha256,
    ).hexdigest()

    return f"sha256={signature}", timestamp


def verify_signature(
    payload: str | bytes,
    signature: str,
    secret: str,
    timestamp: int,
    tolerance_seconds: int = 300,
) -> bool:
    """
    Verify HMAC-SHA256 signature for webhook payload.

    Args:
        payload: The raw JSON payload (string or bytes)
        signature: The signature from X-Webhook-Signature header
        secret: The webhook secret key
        timestamp: The timestamp from X-Webhook-Timestamp header
        tolerance_seconds: Maximum age of request in seconds (default 5 minutes)

    Returns:
        True if signature is valid

    Raises:
        InvalidSignatureError: If signature is invalid or request is too old
    """
    # Check timestamp tolerance
    current_time = int(time.time())
    if abs(current_time - timestamp) > tolerance_seconds:
        raise InvalidSignatureError("Request timestamp is too old or in the future")

    if isinstance(payload, str):
        payload = payload.encode("utf-8")

    # Recreate the signed payload
    signed_payload = f"{timestamp}.".encode("utf-8") + payload

    # Generate expected signature
    expected = hmac.new(
        secret.encode("utf-8"),
        signed_payload,
        hashlib.sha256,
    ).hexdigest()
    expected_signature = f"sha256={expected}"

    # Constant-time comparison to prevent timing attacks
    if not hmac.compare_digest(signature, expected_signature):
        raise InvalidSignatureError("Signature mismatch")

    return True


def parse_signature_header(header: str) -> str:
    """
    Parse the X-Webhook-Signature header value.

    Args:
        header: The header value (e.g., "sha256=abc123...")

    Returns:
        The signature value
    """
    if header.startswith("sha256="):
        return header
    return f"sha256={header}"

File: packages/node/src/webhooks/signature.ts
/**
 * Webhook signature utilities.
 */

import { createHmac, timingSafeEqual } from 'crypto';
import { InvalidSignatureError } from './errors';

/**
 * Generate HMAC-SHA256 signature for webhook payload.
 *
 * @param payload - The raw JSON payload (string or Buffer)
 * @param secret - The webhook secret key
 * @param timestamp - Optional Unix timestamp (defaults to current time)
 * @returns Tuple of [signature, timestamp]
 */
export function generateSignature(
  payload: string | Buffer,
  secret: string,
  timestamp?: number
): [string, number] {
  const ts = timestamp ?? Math.floor(Date.now() / 1000);

  const payloadBuffer =
    typeof payload === 'string' ? Buffer.from(payload, 'utf-8') : payload;

  // Create signed payload: timestamp.payload
  const signedPayload = Buffer.concat([
    Buffer.from(`${ts}.`, 'utf-8'),
    payloadBuffer,
  ]);

  // Generate HMAC-SHA256 signature
  const hmac = createHmac('sha256', secret);
  hmac.update(signedPayload);
  const signature = hmac.digest('hex');

  return [`sha256=${signature}`, ts];
}

/**
 * Verify HMAC-SHA256 signature for webhook payload.
 *
 * @param payload - The raw JSON payload (string or Buffer)
 * @param signature - The signature from X-Webhook-Signature header
 * @param secret - The webhook secret key
 * @param timestamp - The timestamp from X-Webhook-Timestamp header
 * @param toleranceSeconds - Maximum age of request in seconds (default 5 minutes)
 * @returns True if signature is valid
 * @throws InvalidSignatureError if signature is invalid or request is too old
 */
export function verifySignature(
  payload: string | Buffer,
  signature: string,
  secret: string,
  timestamp: number,
  toleranceSeconds: number = 300
): boolean {
  // Check timestamp tolerance
  const currentTime = Math.floor(Date.now() / 1000);
  if (Math.abs(currentTime - timestamp) > toleranceSeconds) {
    throw new InvalidSignatureError(
      'Request timestamp is too old or in the future'
    );
  }

  const payloadBuffer =
    typeof payload === 'string' ? Buffer.from(payload, 'utf-8') : payload;

  // Recreate the signed payload
  const signedPayload = Buffer.concat([
    Buffer.from(`${timestamp}.`, 'utf-8'),
    payloadBuffer,
  ]);

  // Generate expected signature
  const hmac = createHmac('sha256', secret);
  hmac.update(signedPayload);
  const expectedSignature = `sha256=${hmac.digest('hex')}`;

  // Constant-time comparison to prevent timing attacks
  const sigBuffer = Buffer.from(signature, 'utf-8');
  const expectedBuffer = Buffer.from(expectedSignature, 'utf-8');

  if (sigBuffer.length !== expectedBuffer.length) {
    throw new InvalidSignatureError('Signature mismatch');
  }

  if (!timingSafeEqual(sigBuffer, expectedBuffer)) {
    throw new InvalidSignatureError('Signature mismatch');
  }

  return true;
}

/**
 * Parse the X-Webhook-Signature header value.
 *
 * @param header - The header value (e.g., "sha256=abc123...")
 * @returns The signature value
 */
export function parseSignatureHeader(header: string): string {
  if (header.startsWith('sha256=')) {
    return header;
  }
  return `sha256=${header}`;
}
